### PR TITLE
backend unit tests for nml parser and writer

### LINKED ISSUE #2826: backend nml writer + parser unit test

### PR DESCRIPTION
**Pull request by youri-k**
- backend now has unit-tests for nml parser and writer
- they work similar to the frontend tests

### Steps to test:
- call `sbt "test-only *NMLUnitTestSuite"` in wK directory to see the tests running

### Issues:
- fixes #2826 

------
- [x] Ready for review

### CODE REVIEW

**Comment by youri-k** [comment]
@jstriebel these tests currently run in the `e2e-block` of tests. Could you configure that differently?

**Comment by jstriebel** [comment]
@youri-k, do you mean the docker-compose service `e2e-tests`? I can split that into `e2e-tests` and `backend-tests`, yes. The unit tests don't need fossildb or postgres, right?

**Comment by youri-k** [comment]
Right
Thanks :slightly_smiling_face:

File: app/models/annotation/nml/NmlParser.scala [hunk 1]
```diff
@@ -140,7 +140,7 @@ object NmlParser extends LazyLogging with ProtoGeometryImplicits {
   }
 
   private def parseEditRotation(node: NodeSeq) = {
-    node.headOption.flatMap(parseRotation)
+    node.headOption.flatMap(parseRotationForParams)
```
Comment (jstriebel) [inline app/models/annotation/nml/NmlParser.scala]: I guess this fixes a bug you found with the tests?

File: test/backend/NMLUnitTestSuite.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import scala.concurrent.duration.Duration
+
+
+class NMLUnitTestSuite extends FlatSpec with LazyLogging {
+  logger.debug(s"test run")
+  val timestamp = 123456789
+
+  def getObjectId = ObjectId.fromBsonId(BSONObjectID.generate)
+
+  def getParsedTracing(tracing: SkeletonTracing) = {
```
Comment (jstriebel) [inline test/backend/NMLUnitTestSuite.scala]: I would rather rename this to something more active, as it is not a typical `getter`. Maybe `writeAndParseTracing`.

File: test/backend/NMLUnitTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        case (Left(tracing), _) => {
+          assert(tracing == dummyTracing)
+        }
+        case _ => throw new Exception
+      }
+      case _ => throw new Exception
+    }
+  }
+
+  "NML Parser" should "throw an error for invalid comment state" in {
```
Comment (jstriebel) [inline test/backend/NMLUnitTestSuite.scala]: Please add why this is invalid here.

File: test/backend/NMLUnitTestSuite.scala [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    val wrongTree = dummyTracing.trees(1).copy(comments = Seq(Comment(99, "test")))
+    val newTracing = dummyTracing.copy(trees = Seq(dummyTracing.trees(0), wrongTree))
+
+    assert(true)//!isParseSuccessful(getParsedTracing(newTracing)))
+    //TODO: The parser currently doesn't check this
+  }
+
+  it should "throw an error for invalid branchPoint state" in {
```
Comment (jstriebel) [inline test/backend/NMLUnitTestSuite.scala]: Please add why this is invalid here.

File: test/backend/NMLUnitTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    val wrongTree = dummyTracing.trees(1).copy(branchPoints = Seq(BranchPoint(99, 0)))
+    val newTracing = dummyTracing.copy(trees = Seq(dummyTracing.trees(0), wrongTree))
+
+    assert(true)//!isParseSuccessful(getParsedTracing(newTracing)))
+    //TODO: The parser currently doesn't check this
+  }
+
+  it should "throw an error for invalid edge state" in {
```
Comment (jstriebel) [inline test/backend/NMLUnitTestSuite.scala]: Same here

File: app/models/annotation/nml/NmlParser.scala (hunk 1 above)
Comment (youri-k) [reply app/models/annotation/nml/NmlParser.scala]: Yes it is because of difference in parsing `EditRotation` of `nodes` and the general param

**Pull request merged**